File: src/vppgen.cpp
#include "../inc/vppgen.hpp"
#include "../inc/vppconsts.hpp"
#include "../inc/vppstruct.hpp"
#include <string>
#include <vector>

//split the line on the deliminator, ignoring whitespaces
//returns array of split line
void splitLine(const std::string sLine, const char sDelim,
  std::vector<std::vectors<std::string>> &slOut) {
  size_t start;
  size_t end = 0;

  while ((start = str.find_first_not_of(delim, end)) != std::string::npos)
  {
      end = str.find(delim, start);
      slOut.push_back(str.substr(start, end - start));
  }

  //trim white space
  for (size_t i = 0; i < slLine.length(); i++) {
    slOut[i] = slOut[i].trim();
  }
}

//generate the template that gets posted
std::vector<std::string> buildTemplate(const VppInternal &data){
  std::vector<std::string> vRet;

  //species
  if(data.vTagTeam)
    vRet.push_back(std::format(sPkmResTT, data.sBase, data.sName, data.iPost,
      data.vTagTeam.sName, data.vTagTeam.iPost));
  else
    vRet.push_back(std::format(sPkmRes, data.sBase, data.iPost));

  //hatch
  if(data.vTagTeam)
    vRet.push_back(std::format(sHATCHTT, data.sName, data.iPost + data.iHatch,
      data.vTagTeam.sName, data.vTagTeam.iPost + data.vTagTeam.iHatch));
  else
    vRet.push_back(std::format(sHATCH, data.iPost + data.iHatch));

  //complete
  if(data.sSpecies.compare(data.sBase)==0){
    if(data.vTagTeam)
      vRet.push_back(std::format(sLvlMaxSingleTT, data.sName, data.iPost + data.iPostsDone,
        data.vTagTeam.sName, data.vTagTeam.iPost + data.vTagTeam.iPostsDone));
    else
      vRet.push_back(std::format(sLvlMaxSingle, data.iPost + data.iPostsDone));
  }
  else{
    if(data.vTagTeam)
      vRet.push_back(std::format(sLvlMaxEvoTT, data.sSpecies, data.sName, data.iPost + data.iPostsDone,
        data.vTagTeam.sName, data.vTagTeam.iPost + data.vTagTeam.iPostsDone));
    else
      vRet.push_back(std::format(sLvlMaxEvo, data.sSpecies, data.iPost + data.iPostsDone));
  }

  //shiny
  vRet.push_back(std::format(sSHINY, data.bShiny ? sYES : sNO));

  //points
  vRet.push_back(std::format(sPOINTS, data.iPoints));
}

void fillEntry(std::vector<std::string> &slLine, VppInternal &stRet){
  //TODO
}

//build up the structure
//return structure with sorted data
VppInternal fillStruct(const std::vector<std::string> slInput,
  bool bClub = false) {
  VppInternal stRet;
  stRet.bClub = bClub;

  //split all the values of the lines
  std::vector<std::vector<std::string>> slLine;
  for (var i = 0; i < slInput.length(); i++) {
    splitLine(slInput[i], stDelim, slLine);
  }

  //fill in the right property from request
  for(size_t i=0; i<slLine.length(); i++){
    fillEntry(slLine, stRet);
  }

  //TODO: fill in from files


  return stRet;
}

std::vector<std::string> parseAll(const std::vector<std::string> &slInputLeft,
  const std::vector<std::string> &slInputRight, bool bClub){
  //get the requests as structures
  VppInternal stLeftSet = FillStruct(slInputLeft);
  VppInternal stRightSet = FillStruct(slInputRight);

  //tag team?
  if(stLeftSet.sTagTeam.length()>0 && slInputLeft.sTTeam.equals(stRightSet.sName)){
    stLeftSet.vTagTeam = &stRightSet;
    stLeftSet.bClub = false;
  }

  //build output
  std::vector<std::string> vRet = buildTemplate(stLeftSet);

  //return output
  return vRet;
}

File: src/vppgen.pas
unit vppgen;

{$mode DELPHI}{$H+}

interface

uses
  Classes;

function parseAll(const slInputLeft, slInputRight: TStringList;
  bClub: Boolean): TStringList;

var
  slDebug: TStringList;

implementation

uses
  SysUtils, vppconsts, vppstruct;

procedure dbMessage(const sMsg: string);
begin
  slDebug.Add(format('%s %s', [sDebugIdent, sMsg]));
end;

//split the line on the deliminator, ignoring whitespaces
//returns array of split line
procedure splitLine(const sLine, sDelim: string;
  const slOut: TStringList);
var
  i: Integer;
  a: TStringArray;
begin
  slOut.Clear;
  a := sLine.Split(sDelim);

  for i := 0 to High(a) do
    slOut.Add(a[i]);

  for i := 0 to slOut.Count-1 do
    slOut[i] := Trim(slOut[i]);
end;

//generate the template that gets posted
function buildTemplate(const data: TVppInternal): TStringList;
begin
  //species
  result := TStringList.Create;

  if data.vTagTeam<>nil then
    result.Add(Format(sPkmResTT,
      [data.sBase, data.sName, data.iPost,
       data.vTagTeam^.sName, data.vTagTeam^.iPost]))
  else
    result.Add(Format(sPkmRes, [data.sBase, data.iPost]));

  //hatch
  if data.vTagTeam<>nil then
    result.Add(Format(sHATCHTT,
      [data.sName, data.iPost + data.iHatch,
       data.vTagTeam^.sName, data.vTagTeam^.iPost + data.vTagTeam^.iHatch]))
  else
    result.Add(Format(sHATCH, [data.iPost + data.iHatch]));

  //complete
  if data.sSpecies=data.sBase then
  begin
    if data.vTagTeam<>nil then
      result.Add(Format(sLvlMaxSingleTT, [data.sName,
        data.iPost + data.iPostsDone,
        data.vTagTeam^.sName, data.vTagTeam^.iPost +
          data.vTagTeam^.iPostsDone]))
    else
      result.Add(Format(sLvlMaxSingle,
        [data.iPost + data.iPostsDone]));
  end
  else
  begin
    if data.vTagTeam<>nil then
      result.Add(Format(sLvlMaxEvoTT,
        [data.sSpecies, data.sName, data.iPost + data.iPostsDone,
         data.vTagTeam^.sName,
         data.vTagTeam^.iPost + data.vTagTeam^.iPostsDone]))
    else
      result.Add(Format(sLvlMaxEvo, [data.sSpecies,
        data.iPost + data.iPostsDone]));
  end;

  //shiny
  if data.bShiny then
    result.Add(Format(sSHINY, [sYES]))
  else
    result.Add(Format(sSHINY, [sNO]));

  //points
  result.Add(format(sPOINTS, [data.iPoints]));
end;

procedure fillEntry(const slLine: TStringList; var stRet: TVppInternal);
  function IsEqual(const s1, s2: string): Boolean;
  begin
    result := lowercase(s1)=lowercase(s2);
  end;

var
  s: string;
begin
  if slLine.Count<2 then
    Exit;

  //Name
  for s in slName do
    if IsEqual(slLine[0], s) then
    begin
      stRet.sName:=slLine[1];
      break;
    end;

  //Species
  for s in slPkm do
    if IsEqual(slLine[0], s) then
    begin
      stRet.sSpecies:=slLine[1];
      break;
    end;

  //Tag Team
  for s in slPkm do
    if IsEqual(slLine[0], s) then
    begin
      stRet.sTTeam:=slLine[1];
      break;
    end;

  //PostCount
  for s in slPkm do
    if IsEqual(slLine[0], s) then
    begin
      slLine[1] := StringReplace(slLine[1], ',', '', []);
      slLine[1] := StringReplace(slLine[1], '.', '', []);
      stRet.iPost:=StrToInt(slLine[1]);
      break;
    end;

  //Hemisphere
  for s in slPkm do
    if IsEqual(slLine[0], s) then
    begin
      stRet.sHemi:=slLine[1];
      break;
    end;

  //Rainbow Stone
  for s in slPkm do
    if IsEqual(slLine[0], s) then
    begin
      //some people put a '!' at the end so we just look if there is a 'yes'
      stRet.bStone:=AnsiPos(lowercase(sYES), lowercase(slLine[1])>0;
      break;
    end;
end;

//build up the structure
//return structure with sorted data
function fillStruct(const slInput: TStringList;
  bClub: Boolean = false): TVppInternal;
var
  stRet: TVppInternal;
  slLines: TStringList;
  i: Integer;
begin
  stRet.bClub := bClub;
  slLines := TStringList.Create;
  try
    //split all the values of the lines
    for i := 0 to slInput.Count-1 do
    begin
      slLines.AddObject('', TStringList.Create);
      splitLine(slInput[i], sDelim, slLines.Objects[i] as TStringList);
    end;

    //fill in the right property from request
    for i := 0 to slLines.Count-1 do
      fillEntry(slLines.Objects[i] as TStringList, stRet);

    //TODO: fill in from json files



  finally
    for i := 0 to slLines.Count-1 do
      (slLines.Objects[i] as TStringList).Free;
  end;
  result := stRet;
end;

function parseAll(const slInputLeft, slInputRight: TStringList;
  bClub: Boolean): TStringList;
var
  stLeftSet, stRightSet: TVppInternal;
begin
  slDebug.Clear;
  dbMessage('---start debug information---');

  //get the requests as structures
  stLeftSet := FillStruct(slInputLeft, bClub);
  stRightSet := FillStruct(slInputRight, bClub);

  //tag team?
  if (Length(stLeftSet.sTTeam)>0) and
    (stLeftSet.sTTeam=stRightSet.sName) then
  begin
    stLeftSet.vTagTeam := @stRightSet;
    stLeftSet.bClub := false;
    dbMessage('club membership ignored!');
  end;

  //build output
  result := buildTemplate(stLeftSet);

  dbMessage('---end debug information---');
end;

initialization
  slDebug := TStringList.Create;
finalization
  FreeAndNil(slDebug);

end.
